File: archivist/src/test/java/com/zorroa/archivist/web/JobControllerTests.java
package com.zorroa.archivist.web;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import com.zorroa.archivist.domain.*;
import com.zorroa.archivist.service.JobService;
import com.zorroa.common.domain.TaskState;
import com.zorroa.sdk.domain.PagedList;
import com.zorroa.sdk.domain.Pager;
import com.zorroa.sdk.processor.ProcessorRef;
import com.zorroa.sdk.search.AssetSearch;
import com.zorroa.sdk.util.Json;
import com.zorroa.sdk.zps.ZpsScript;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockHttpSession;
import org.springframework.test.web.servlet.MvcResult;

import static org.junit.Assert.assertEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

/**
 * Created by chambers on 10/4/16.
 */
public class JobControllerTests extends MockMvcTest {

    @Autowired
    JobService jobService;

    JobSpec spec;
    Job job;
    TaskSpec tspec;
    Task task;

    @Before
    public void init() {
        JobSpec spec = new JobSpec();
        spec.setName("foo");
        spec.setType(PipelineType.Export);
        job = jobService.launch(spec);

        tspec = new TaskSpec();
        tspec.setName("task1");
        tspec.setScript(new ZpsScript());
        tspec.setJobId(job.getJobId());
        task = jobService.createTask(tspec);

        job = jobService.get(job.getId());
    }

    @Test
    public void testLaunch() throws Exception {
        MockHttpSession session = admin();

        ZpsScript script = new ZpsScript();
        script.setGenerate(ImmutableList.of(
                new ProcessorRef("com.zorroa.core.generator.AssetSearchGenerator")
                        .setArg("search", new AssetSearch("test"))));

        script.setExecute(ImmutableList.of(
                new ProcessorRef("com.zorroa.core.processor.GroupProcessor")));

        JobSpecV spec = new JobSpecV();
        spec.setScript(script);
        spec.setName("Test job");
        spec.setArgs(ImmutableMap.of("arg1", "value1"));
        spec.setEnv(ImmutableMap.of("env1", "value1"));
        spec.setType(PipelineType.Import);

        MvcResult result = mvc.perform(post("/api/v1/jobs")
                .session(session)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(Json.serialize(spec)))
                .andExpect(status().isOk())
                .andReturn();

        Job job1 = deserialize(result, Job.class);
        Job job2 = jobService.get(job1.getId());
        assertEquals(job1, job2);
        assertEquals(1, job2.getCounts().getTasksTotal());

        PagedList<Task> tasks = jobService.getAllTasks(job2.getJobId(), Pager.first());
        assertEquals(1, tasks.size());
    }

    @Test(expected=org.springframework.web.util.NestedServletException.class)
    public void testLaunchValidationFailure() throws Exception {
        MockHttpSession session = admin();

        JobSpecV spec = new JobSpecV();
        mvc.perform(post("/api/v1/jobs")
                .session(session)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .content(Json.serialize(spec)))
                .andExpect(status().isOk())
                .andReturn();
    }

    @Test
    public void getTasksByJob() throws Exception {
        MockHttpSession session = admin();

        MvcResult result = mvc.perform(get("/api/v1/jobs/" + job.getId() + "/tasks")
                .session(session)
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        PagedList<Task> tasks = deserialize(result, new TypeReference<PagedList<Task>>() {});
        assertEquals(1, tasks.size());
    }

    @Test
    public void getTasks() throws Exception {
        MockHttpSession session = admin();

        TaskFilter f = new TaskFilter();
        f.setStates(ImmutableSet.of(TaskState.Waiting));
        f.setSort(ImmutableMap.of("taskId", "desc"));

        MvcResult result = mvc.perform(post("/api/v2/jobs/" + job.getId() + "/tasks")
                .session(session)
                .content(Json.serialize(f))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(status().isOk())
                .andReturn();

        PagedList<Task> tasks = deserialize(result, new TypeReference<PagedList<Task>>() {});
        assertEquals(1, tasks.size());
    }

}

File: archivist/src/test/kotlin/com/zorroa/archivist/web/JobControllerTests.kt
package com.zorroa.archivist.web

import com.zorroa.archivist.domain.emptyZpsScript
import com.zorroa.archivist.service.JobService
import com.zorroa.common.domain.*
import com.zorroa.common.util.Json
import org.junit.Before
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class JobControllerTests : MockMvcTest() {

    @Autowired
    lateinit var jobService: JobService

    lateinit var job: Job

    @Before
    fun init() {
        val spec = JobSpec("test_job",
                emptyZpsScript("foo"),
                args=mutableMapOf("foo" to 1),
                env=mutableMapOf("foo" to "bar"))

        job = jobService.create(spec)
    }

    @Test
    fun testGet() {
        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.get("/api/v1/jobs/" + job.id)
                .session(session)
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val t1 = deserialize(result, Job::class.java)
        assertEquals(job.id, t1.id)
    }

    @Test
    fun testCreate() {
        val spec = JobSpec("test_job_2",
                emptyZpsScript("test"),
                args=mutableMapOf("foo" to 1),
                env=mutableMapOf("foo" to "bar"))

        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.post("/api/v1/jobs")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .content(Json.serialize(spec))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val t1 = deserialize(result, Job::class.java)
        assertEquals(spec.name, t1.name)
        assertEquals(1, t1.taskCounts!!["tasksTotal"])
        assertEquals(1, t1.taskCounts!!["tasksWaiting"])
    }

    @Test
    fun testUpdate() {
        val spec = JobUpdateSpec(name="silly_bazilly", priority = 5)

        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.put("/api/v1/jobs/${job.id}")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .content(Json.serialize(spec))
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val t1 = deserialize(result, Job::class.java)
        assertEquals(spec.name, t1.name)
        assertEquals(spec.priority, t1.priority)
    }

    @Test
    fun testCancel() {
        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.put("/api/v1/jobs/${job.id}/_cancel")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val job = jobService.get(job.id)
        assertEquals(JobState.Cancelled, job.state)

        val status = deserialize(result, Json.GENERIC_MAP)
        assertEquals("Job", status["type"])
        assertEquals("cancel", status["op"])
        assertEquals(true, status["success"])
    }

    @Test
    fun testRestart() {
        jobService.setJobState(job, JobState.Cancelled, null)
        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.put("/api/v1/jobs/${job.id}/_restart")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val job = jobService.get(job.id)
        assertEquals(JobState.Active, job.state)

        val status = deserialize(result, Json.GENERIC_MAP)
        assertEquals("Job", status["type"])
        assertEquals("restart", status["op"])
        assertEquals(true, status["success"])
    }

    @Test
    fun testRetryAllFailures() {
        val t = jobService.createTask(job, TaskSpec("foo", emptyZpsScript("bar")))
        assertTrue(jobService.setTaskState(t, TaskState.Failure, null))

        val session = admin()
        val result = mvc.perform(MockMvcRequestBuilders.put("/api/v1/jobs/${job.id}/_retryAllFailures")
                .session(session)
                .with(SecurityMockMvcRequestPostProcessors.csrf())
                .contentType(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andReturn()

        val status = deserialize(result, Json.GENERIC_MAP)
        assertEquals("Job", status["type"])
        assertEquals("retryAllFailures", status["op"])
        assertEquals(true, status["success"])

        val t2 = jobService.getTask(t.id)
        assertEquals(TaskState.Waiting, t2.state)
    }
}